### PR TITLE
Frame graph: refactor to use InternalTexture instead of RenderTargetWrapper + misc changes

### PR DESCRIPTION
**Pull request by Popov72**
I've made a few changes so that I can create an empty `RenderTargetWrapper` and “connect” the render textures + depth texture later:
* New `dontCreateTextures` options in `IMultiRenderTargetOptions`. This way, `Engine.createMultipleRenderTarget` will not create textures (color/depth).
* New `createMSAATexture` options in `InternalTextureCreationOptions`. This option allows `Engine._createInternalTexture` to immediately create the MSAA texture (`WebGLRenderBuffer` in WebGL) associated with the internal texture. For the moment, these textures are only created when you call `updateRenderTargetTextureSampleCount` or `updateMultipleRenderTargetTextureSampleCount`, in conjunction with a `RenderTargetWrapper`. Note that this option is not used in WebGPU, as this is already the way it works now.
* New setter `RenderTargetWrapper.depthStencilTexture` (we already have a getter), to be able to define the depth/stencil texture of a `RenderTargetWrapper` from an existing texture.
* `Engine._createInternalTexture` is now able to create a depth/stencil texture. Previously, you could only create one by calling `Engine._createDepthStencilTexture`, and it was automatically associated with a `RenderTargetWrapper`.

With these changes, it's easy to create all color/depth textures by calling `Engine._createInternalTexture`, and combine them as you wish in a `RenderTargetWrapper` to create a render target (this is what a frame graph render pass does - see the new `FrameGraphRenderTarget` class).

Other changes:
* I've put `IsDepthTexture`, `GetTypeForDepthTexture` and `HasStencilAspect` in `internalTexture.ts`. I don't know if there's a better place... I tried to put them in `Misc/textureTools.ts` first, but there was a problem with the imports because I need to use these functions in `thinEngine.ts`.
* I removed the functions in `FrameGraph.ts` that only redirected to the corresponding functions in `FrameGraphTextureManager`.
* The NRG `Clear` block can now clear a color and depth texture at the same time.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15874/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15874/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15874/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15874/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15874/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15874/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15874/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15874/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15874/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15874/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts [hunk 1]
```diff
@@ -43,6 +45,38 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      */
     public _currentLOD = 0;
 
+    public override get depthStencilTexture() {
+        return this._depthStencilTexture;
```
Comment (sebavan) [inline packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:49]: The override does not seem to change anything ?

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts [hunk 2]
```diff
@@ -43,6 +45,38 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      */
     public _currentLOD = 0;
 
+    public override get depthStencilTexture() {
+        return this._depthStencilTexture;
+    }
+
+    public override set depthStencilTexture(texture: Nullable<InternalTexture>) {
```
Comment (sebavan) [inline packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:52]: is the override keyword needed if the setter is not in the base class ?

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts [hunk 3]
```diff
@@ -142,7 +176,8 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      * @param lodLevel defines the lod level to bind to the frame buffer
      */
     private _bindTextureRenderTarget(texture: InternalTexture, attachmentIndex: number = 0, faceIndexOrLayer?: number, lodLevel: number = 0) {
-        if (!texture._hardwareTexture) {
+        const hardwareTexture = texture._hardwareTexture as WebGLHardwareTexture;
```
Comment (sebavan) [inline packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:179]: Not sure why the change is necessary ?

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 4]
```diff
@@ -2841,20 +2841,26 @@ export class ThinEngine extends AbstractEngine {
...
         let samplingMode = Constants.TEXTURE_TRILINEAR_SAMPLINGMODE;
         let format = Constants.TEXTUREFORMAT_RGBA;
         let useSRGBBuffer = false;
         let samples = 1;
         let label: string | undefined;
+        let createMSAATexture = false;
+        let comparisonFunction = 0;
         if (options !== undefined && typeof options === "object") {
             generateMipMaps = !!options.generateMipMaps;
+            createMipMaps = !!options.createMipMaps;
```
Comment (sebavan) [inline packages/dev/core/src/Engines/thinEngine.ts:2855]: What is the difference with generateMipMaps ?

File: packages/dev/core/src/Engines/thinEngine.ts [hunk 5]
```diff
@@ -3468,7 +3547,8 @@ export class ThinEngine extends AbstractEngine {
         width: number,
         height: number,
         samples = 1,
-        depthTextureFormat?: number
+        depthTextureFormat?: number,
+        dontBindRenderBufferToFrameBuffer = false
```
Comment (sebavan) [inline packages/dev/core/src/Engines/thinEngine.ts:3551]: Not sure to understand when this would be used ? (creating the buffers but not binding them)

**Review by sebavan — COMMENTED** [review]
As you are switching is there any work done in the past on the Wrapper that should be removed ?

**Comment by Popov72** [comment]
> As you are switching is there any work done in the past on the Wrapper that should be removed ?

I don't think so:

https://github.com/BabylonJS/Babylon.js/pull/15735/files#diff-1c9b50019ad821feea698fd26f1616e22fd9b1e65a920cdbd70c03047428a716

Looking at my other changes in engine.multiRender.ts / engine.renderTarget.ts, I think we can keep all of them.

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts (hunk 1 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:49]: I though we had to override both the getter and setter when we override one of them, but it's not the case. I will remove the getter.

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts (hunk 2 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:52]: The setter (as well as the getter) is defined in the base class (`RenderTargetWrapper`).

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts (hunk 3 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:179]: Just because `texture._hardwareTexture` is used several times (5) in the function.

File: packages/dev/core/src/Engines/thinEngine.ts (hunk 4 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/thinEngine.ts:2855]: `createMipMaps` is used to create additional storage for the mipmaps when the texture is created, while `generateMipMaps` is used to indicate that we should generate the mipmaps when we unbind a MSAA framebuffer. It's possible to have `createMipMaps=true` and `generateMipMaps=false`, which means that the texture should be created with storage for mipmaps, but that mipmaps generation will be handled by the user and not automatically.

File: packages/dev/core/src/Engines/thinEngine.ts (hunk 5 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/thinEngine.ts:3551]: It's used in `engine._createInternalTexture`, when creating the MSAA texture (=`WebGLRenderBuffer`). In that case, the MSAA framebuffer does not exist yet (this framebuffer is managed by `WebGLRenderTargetWrapper`), so we can't bind the render buffer to it. Binding the MSAA render buffer is managed by the new code in `WebGLRenderTargetWrapper._bindTextureRenderTarget` (at the end of the method).

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts (hunk 2 above)
Comment (sebavan) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:48]: was looking locally...

**Comment by sebavan** [comment]
you can merge master back in for the flaky test

**Comment by Popov72** [comment]
> you can merge master back in for the flaky test

In fact I do have a problem with a test with my PR, I must fix it...

**Comment by Popov72** [comment]
It seems I was right, after all, if you override a setter, you must also override the getter, else you get "undefined" from the getter (see https://stackoverflow.com/questions/28950760/override-a-setter-and-the-getter-must-also-be-overridden for eg)! That was the reason for the fluid rendering visu test to fail.

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts [hunk 6]
```diff
@@ -43,6 +45,38 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      */
     public _currentLOD = 0;
 
+    public override get depthStencilTexture() {
+        return this._depthStencilTexture;
+    }
+
+    public override set depthStencilTexture(texture: Nullable<InternalTexture>) {
+        this._depthStencilTexture = texture;
```
Comment (sebavan) [inline packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:53]: Couldnt this leak ? and should we dispose the previous texture ?

**Review by sebavan — COMMENTED** [review]
Last comment and LGTM

File: packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts (hunk 6 above)
Comment (Popov72) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:53]: I don't dispose the existing texture if it exists, because in the frame graph, the texture lifetime is handled by the frame graph texture manager. Maybe I should use a regular function instead of a setter and pass a "disposeExisting" boolean so that the default behavior is to dispose the existing texture?

Comment (sebavan) [reply packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:53]: would be safer indeed as I am pretty sure a misuse of this will be hard to find in the future :-)

**Review by sebavan — APPROVED** [review]

**Review by sebavan — APPROVED** [review]

**Comment by Jeggery** [comment]
Is this change stable? It has too much impact and I dare not update the latest version anymore

**Comment by sebavan** [comment]
We would have not merged it if unstable, and if smthg was wrong I bet the community would have caught it already :-)
